src/main/java/org/radarcns/management/web/rest/MetaTokenResource.java

```diff
@@ -0,0 +1,57 @@
+package org.radarcns.management.web.rest;
+
+
+import java.net.MalformedURLException;
+import java.time.Duration;
+import java.util.concurrent.TimeUnit;
+
+import com.codahale.metrics.annotation.Timed;
+import org.radarcns.auth.config.Constants;
+import org.radarcns.management.service.MetaTokenService;
+import org.radarcns.management.service.dto.ClientPairInfoDTO;
+import org.radarcns.management.service.dto.TokenDTO;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/api")
+public class MetaTokenResource {
+
+    private final Logger log = LoggerFactory.getLogger(OAuthClientsResource.class);
+
+    public static final Duration DEFAULT_META_TOKEN_TIMEOUT = Duration.ofHours(1);
+
+    @Autowired
+    private MetaTokenService metaTokenService;
+
+    /**
+     * GET /api/meta-token/:tokenName.
+     *
+     * <p>Get refresh-token available under this tokenName.</p>
+     *
+     * @param tokenName the tokenName given after pairing the subject with client
+     * @return the client as a {@link ClientPairInfoDTO}
+     */
+    @GetMapping("/meta-token/{tokenName:" + Constants.TOKEN_NAME_REGEX + "}")
+    @Timed
+    public ResponseEntity<TokenDTO> getTokenByTokenName(@PathVariable("tokenName") String tokenName)
+            throws MalformedURLException {
+        log.info("Requesting token with tokenName {}", tokenName);
+        try {
+            // sleep for 5 seconds
+            TimeUnit.SECONDS.sleep(5);
```

Comment:
Why do we need to `sleep()` here?

Reply:
Perhaps 5 seconds is excessive (3 seconds would probably do), but it makes brute-force attacks too expensive to be worthwhile. For every guess, an attacker needs to wait X seconds before a response is returned. This needs to be done after both valid and invalid requests, to ensure that an attacker does not define an internal time threshold that would only be reached for invalid URLs.

Attack pseudo-code:
```python
for sequence in sequences_to_try():
     response = request(sequence, timeout_ms=500)
     if response.successful:
         # keys to the castle!
```

Perhaps it would be good to mention this in the comments, indeed.

Reply:
I checked with gatling to be sure, and it is as i suspected. This does indeed make each request take a long time, but requests are served asynchronously. So an attacker can just make 100's of requests per second, and would receive all the responses a few seconds later.

See this image, where we can see around 15 requests per second, and also 15 responses per second, as opposed to one response per 5 seconds. Generated using [this](https://gist.github.com/dennyverbeeck/79836a3a529154ebfca6d38268e2d5ba) code.
![image](https://user-images.githubusercontent.com/10878787/44456564-258d5e80-a601-11e8-90a6-57afd9859016.png)

We will probably need to add an interceptor to implement rate limiting based on the client's IP address. I suggest we remove the sleep and move the rate limiting implementation to a new issue.

Reply:
There is a limit in the number of simultaneous connections for clients though, which makes the sleep more effective. I agree that a rate limiting implementation would be preferable, but I would rather wait for that implementation and then remove the sleep. In the context of RADAR, perhaps the rate limiting needs to happen on the nginx level, since that is the reverse proxy on top of MP.

Reply:
I agree to remove the sleep after we have the rate-limiting. IMO, having the sleep doesn't do any harm.